src/core/lwt.ml

```diff
@@ -51,7 +378,45 @@ struct
     | Pending      : 'a callbacks        -> ('a, underlying, pending)   state
     | Unified_with : ('a, _, 'c) promise -> ('a, proxy,      'c)        state
 
+  (* Note:
+
+     A promise whose state is [Unified_with _] is a "proxy" promise. A promise
+     whose state is *not* [Unified_with _] is an "underlying" promise.
 
+     The "underlying promise of [p]" is:
+
+     - [p], if [p] is itself underlying.
+     - Otherwise, [p] is a proxy and has state [Unified_with p']. The underlying
+       promise of [p] is the underlying promise of [p'].
+
+     In other words, to find the underlying promise of a proxy, Lwt follows the
+     [Unified_with _] links to the end. *)
+
+  (* Note:
+
+     When a promise is completed, or becomes a proxy, its state field is
+     mutated. This invalidates the type invariants on the promise. See internal
+     function [set_promise_state] for details about that.
+
+     When an Lwt function has a reference to a promise, and also registers a
+     callback that has a reference to the same promise, the invariants on the
+     reference may become invalid by the time the callback is called. All such
+     callbacks have comments explaining what the valid invariants are at that
+     point, and/or casts to (1) get the correct typing and (2) document the
+     potential state change for readers of the code. *)
+
+  (* Note:
+
+     The phantom types used in the GADT are declared as abstract types. They
+     must be declared outside the module, for OCaml's type system to know that
+     they are definitely distinct types. If they are declared inside this
+     module, OCaml sees them through an inferred signature in the rest of
+     [lwt.ml]. That signature does not guarantee that they are distinct, which
+     defeats the purpose of the GADT in pattern matching. *)
```

Comment:
While this note is partially correct, you can get rid of the problem by giving a constructor for your phantom types (and expose the constructor):
```ocaml
type underlying = Underlying
```

This is a better way to prove to the typechecker that underlying is different to every other types.